### PR TITLE
[FEATURE] Allow expression functions to use named parameters

### PR DESCRIPTION
**Pull request by nyalldawson**
This commit sets the framework for allowing expression functions to use named parameters. Ie, instead of: 

clamp(1,2,3)

you can use:

clamp( min:=1, value:=2, max:=3)

This also allows arguments to be switched, eg:

clamp( value:=2, max:=3, min:=1)

Additionally, it allows for a more structured definition of function parameters to handle optional arguments and default values for parameters. These are currently being done using a hacky infinite argument list.

I've utilised the postgres ':=' syntax for specifying named arguments to avoid potential collisions which may arise with the equality test if we re-used just the '=' operator alone.

This PR is a proof of concept only (eg incomplete docs, tests and bindings), but I'm after feedback on whether this is the cleanest approach to take.

Basically, all the work is done when either preparing an expression or evaluating it for the first time. A map is built up for translating the specified parameters into the order expected by the function, and replacing any missing parameters with their default values.

### CODE REVIEW

**Comment by NathanW2** [comment]
Nice! Was this a  'I'm on holiday but have a cool QGIS idea' one?

**Comment by m-kuhn** [comment]
From an concept perspective, this sounds good to me. :+1:

**Comment by jef-n** [comment]
hm, one step farer away from sql...

**Comment by nyalldawson** [comment]
@jef-n maybe, but postgres supports this syntax ;)

**Comment by m-kuhn** [comment]
and it's an extension that can be mapped to standard sql, not an incompatibility :)

**Comment by nyalldawson** [comment]
Thinking more about this - there's a much cleaner way this can be implemented. I'll leave this open for now for feedback about the syntax and whether or not this feature is desirable.

@jef-n for clarification - are you totally against seeing something like this implemented?

**Comment by jef-n** [comment]
@nyalldawson not really.  I guess this doesn't make is significantly more difficult to map stuff to SQL.  We already have lots of other stuff in expressions anyway.

**Comment by NathanW2** [comment]
This is optional @nyalldawson?  So the first example will still continue to work.

**Comment by nyalldawson** [comment]
Ok... I've updated this with a much improved implementation. Now the parameter handling is done during parsing. When a function with named parameters is parsed, the argument list will be internal rearranged to match the order expected by the function and any non-specified parameters will be replaced by a literal node representing their default value. This approach simplifies the code and also means that expressions with named parameters can still be handled easily by expression compilers.

**Comment by nyalldawson** [comment]
Manually merged
